pallets/pooled-staking/src/pools.rs

```diff
@@ -0,0 +1,393 @@
+        Ok(())
+    }
+
+    fn decrease_hold(
+        candidate: &Candidate<T>,
+        delegator: &Delegator<T>,
+        stake: &Stake<T>,
+    ) -> Result<(), Error<T>> {
+        let hold = Self::hold(candidate, delegator);
+        let hold = hold.0.err_sub(&stake.0)?;
+        Self::set_hold(candidate, delegator, Stake(hold));
+        Ok(())
+    }
+}
+
+pub fn check_candidate_consistency<T: Config>(candidate: &Candidate<T>) -> Result<(), Error<T>> {
+    let total0 = Pools::<T>::get(candidate, &PoolsKey::CandidateTotalStake);
+
+    let joining = Joining::<T>::total_staked(&candidate).0;
+    let auto = AutoCompounding::<T>::total_staked(&candidate).0;
+    let manual = ManualRewards::<T>::total_staked(&candidate).0;
+
+    let total1 = joining
+        .checked_add(&auto)
+        .ok_or(Error::InconsistentState)?
+        .checked_add(&manual)
+        .ok_or(Error::InconsistentState)?;
+
+    ensure!(total0 == total1, Error::InconsistentState);
+
+    Ok(())
+}
+
+macro_rules! impl_pool {
+    ($name:ident, $shares:ident, $supply:ident, $total:ident, $hold: ident, $init:expr $(,)?) => {
+        pub struct $name<T>(PhantomData<T>);
+        impl<T: Config> Pool<T> for $name<T> {
+            fn shares(candidate: &Candidate<T>, delegator: &Delegator<T>) -> Shares<T> {
+                Shares(Pools::<T>::get(
+                    candidate,
+                    &PoolsKey::$shares {
+                        delegator: delegator.clone(),
+                    },
+                ))
+            }
+
+            fn shares_supply(candidate: &Candidate<T>) -> Shares<T> {
+                Shares(Pools::<T>::get(candidate, &PoolsKey::$supply))
+            }
+
+            fn total_staked(candidate: &Candidate<T>) -> Stake<T> {
+                Stake(Pools::<T>::get(candidate, &PoolsKey::$total))
+            }
+
+            fn hold(candidate: &Candidate<T>, delegator: &Delegator<T>) -> Stake<T> {
+                Stake(Pools::<T>::get(
+                    candidate,
+                    &PoolsKey::$hold {
+                        delegator: delegator.clone(),
+                    },
+                ))
+            }
+
+            fn set_shares(candidate: &Candidate<T>, delegator: &Delegator<T>, value: Shares<T>) {
+                Pools::<T>::set(
+                    candidate,
+                    &PoolsKey::$shares {
+                        delegator: delegator.clone(),
+                    },
+                    value.0,
+                )
+            }
+
+            fn set_shares_supply(candidate: &Candidate<T>, value: Shares<T>) {
+                Pools::<T>::set(candidate, &PoolsKey::$supply, value.0)
+            }
+
+            fn set_total_staked(candidate: &Candidate<T>, value: Stake<T>) {
+                Pools::<T>::set(candidate, &PoolsKey::$total, value.0)
+            }
+
+            fn set_hold(candidate: &Candidate<T>, delegator: &Delegator<T>, value: Stake<T>) {
+                Pools::<T>::set(
```

Comment:
For each candidate that joins and then leaves, the storage grows a few bytes. All these functions should kill the storage item when the value is 0. But not sure if that solves the problem because an attacker could simply leave some dust and pollute the storage anyway. Do you think this is worth it?

And does anyone know if this being a `ValueQuery` means that a value of `0` is handled specially? Because it doesn't seem like that in tests, but in production it may be different.